rfcs/0116-artifact-service.md

```diff
@@ -0,0 +1,196 @@
+* We want to have more clear responsibilities for our services
+* We want to consolidate the handling of Artifacts into a single service
+* We want to build generalised software when possible as participants of a free
+  software community
+
+# Details
+
+* The Artifact service will start from the Queue's `artifacts.js` file
+* Auth might not be done with scopes and tc-auth, but a shared secret between
+  Queue and Artifacts
+* The Queue will block new tasks after task resolution by not issuing new
+  signed urls
+* Data storage in Artifacts will be done with Postgres and not Azure Table
+  Storage
+* We do not intend to have an operation to list artifacts in the API
+* The redirecting logic of the Artifact service will adhere to the HTTP spec
+  using 300-series redirects where possible
+* Where content-negotiation is impossible, the service will block creation of
+  artifacts when incoming requests aren't compatible with required response
+* We will support uploading artifacts to different regions so that we can
+
+## API
+This API is up for debate and is only an illustrative starting point.  
+
+### Generalised request format
+There will be a concept of a generalised HTTP request.  This will be an object shaped like
+```javascript
+{
+  method: 'GET',
+  url: 'https://www.example.com',
+  headers: {
+    'user-agent': 'taskcluster',
+  }
+}
+```
+
+* The body will not be included, but can be described through HTTP headers in the request
+* Duplicated HTTP headers will be unsupported (i.e. key -> string not key -> list of strings)
+* All query string, path, fragements are contained in a fully formed `url` property
+
+### Creation
+
+```
+PUT /artifacts/:name
+```
+
+This endpoint will take an origin and name.  It will return a `200` with a
+document which contains a list of generalised requests which could be run to
+complete the actual upload.  Please see the queue's current `blob` storage type
+for an example of this pattern.  The body of the request will contain metadata
+about the artifact in the shape:
+
+```javascript
+{
+  origin: '10.10.10.10' || 's3_us-west-2',
+  contentType: 'application/json',
+  contentLength: 1234,
+  contentSha256: 'abcdef1234',
+  transferLength: 1234,
+  transferSha256: 'abcdef1234',
+  contentEncdoing: 
+  expiration: new Date(),
+}
+```
+
+The `origin` parameter will either be an IPv4 address, IPv6 address, or an
+identifier.  This parameter will specify the source location of the ultimately
+originating request.  This means the origin of the request to the Queue from
+the worker and not the request to the artifact service.  If the origin is an
+IPv4 or IPv6, a mapping of the address and IP will occur to find the backing
+storage.  If the origin is not an IPv4 or IPv6 address and is identical to a
+set of known and mappable identifiers (e.g. `s3_us-west-2`), then that will be
+used to map the token to find the backing storage.
+
+### Deletetion
+
+```
+DELETE /artifacts/:name
+DELETE /caches/artifacts/:name
```

Comment:
Tooltool doesn't support deletion at all, and TC doesn't for artifacts, either, so I tend to agree.  If this is really easy to implement, fine, but if the implementation is complicated or racy then perhaps not.  It will probably be useful once or twice to delete some accidentally-uploaded private data.  That's happened before with tooltool, for example (and required some manual database and S3 operations to clear up).

Reply:
it'll be racy by definition...

As in there is a possibility for racing between creating and deleting the resource, especially considering that uploads can retry.

If we really need to purge a secret that shoudn't have been uploaded, it's better open the database... People could put such a secret in the task definition too... this could always happen.

Reply:
which would you prefer to do if you were the one cleaning up a chemspill level leak:

```
objadm --auth-token abcdef1234 delete our-pgp-private-key.txt
```

or 

1. Figure out how the service stores data
3. Figure out which tools need to be used to access the database
2. Figure out the credentials for the database
52. Edit the database schemas underlying
2. Figure out how to access every underlying object store 
9. Figure out how to map the `:name` to each underlying object store's version of the name
6. Figure out how to delete things safely from all the regions
33. Delete all the files.

By the way: don't delete anything you're not supposed to.  This could be as easy as accidentally forgetting a `WHERE` clause on your `DELETE`.

We wouldn't be exposing this part of the API to anyone other than the Queue and ourselves.  "Just opening the database" is a gross simplification here, and assumes that the people who will be operating this service are knowledgable on Postgres, Node.js, Taskcluster services.  We wouldn't be deleting things often, but it is pretty much expected that if you create and delete the same resource at the same time, you'd have weird behaviour.

As someone who does understand these things, in production, I'd rather use a unit tested and known-to-work option than opening a sql client and the S3 management console.

As for cache clearing, it's by definition and per the HTTP spec that it's very clear to be a best-effort *initiation* of a delete.